src/IECoreDelight/ShaderNetworkAlgo.cpp

```diff
@@ -301,20 +669,242 @@ namespace IECoreDelight
+	const ShaderNetwork *shaderNetwork,
+	USDLightType usdLightType
+)
+{
+	DelightHandleSharedPtr result;
+
+	const std::string lightName = ( "light:" + std::string( transformHandle.name() ) ).c_str();
+	switch( usdLightType )
+	{
+		case USDLightType::SphereLight:
+			{
+				ConstObjectPtr geo = new IECoreScene::SpherePrimitive( g_defaultRadius );
+				NodeAlgo::convert( geo.get(), transformHandle.context(), lightName.c_str() );
+				result = std::make_shared<DelightHandle>( transformHandle.context(), lightName, transformHandle.ownership() );
+			}
+			break;
+		case USDLightType::RectLight:
+			{
+				ConstObjectPtr geo = quad( g_defaultWidth, g_defaultHeight );
+				NodeAlgo::convert( geo.get(), transformHandle.context(), lightName.c_str() );
+				result = std::make_shared<DelightHandle>( transformHandle.context(), lightName, transformHandle.ownership() );
+			}
+			break;
+		case USDLightType::DiskLight:
+			{
+				ConstObjectPtr geo = new DiskPrimitive( g_defaultRadius );
+				NodeAlgo::convert( geo.get(), transformHandle.context(), lightName.c_str() );
+				result = std::make_shared<DelightHandle>( transformHandle.context(), lightName, transformHandle.ownership() );
+			}
+			break;
+		case USDLightType::DistantLight:
+			{
+				result = std::make_shared<DelightHandle>( transformHandle.context(), lightName, transformHandle.ownership(), "environment" );
+			}
+			break;
+		case USDLightType::DomeLight:
+			{
+				ParameterList p;
+				p.add( g_angleParameter.c_str(), new DoubleData( 360 ) );
+				result = std::make_shared<DelightHandle>( transformHandle.context(), lightName, transformHandle.ownership(), "environment", p );
+			}
+			break;
+		case USDLightType::CylinderLight:
+			{
+				ConstObjectPtr geo = cylinder( 0.5f, 1.f );
+				NodeAlgo::convert( geo.get(), transformHandle.context(), lightName.c_str() );
+				result = std::make_shared<DelightHandle>( transformHandle.context(), lightName, transformHandle.ownership() );
+			}
+			break;
+		default: break;
+	}
+
+	return result;
+}
+
+void updateUSDLightGeometry(
+	const DelightHandle &transformHandle,
+	const DelightHandle *geometryHandle,
+	const ShaderNetwork *shaderNetwork,
+	USDLightType usdLightType
+)
+{
+	if( const Shader *light = shaderNetwork->outputShader() )
+	{
+		const std::string lightName = ( "light:" + std::string( transformHandle.name() ) ).c_str();
+		ParameterList parameters;
+		switch( usdLightType )
+		{
+			case USDLightType::SphereLight:
+				{
+					const float width = parameterValue( light, g_radiusParameter, g_defaultRadius ) * 2.f;
+					parameters.add( { "width", &width, NSITypeFloat, 0, 1, 0 } );
+				}
+				break;
+			case USDLightType::RectLight:
+				{
+					const float width = parameterValue( light, g_widthParameter, g_defaultWidth );
+					const float height = parameterValue( light, g_heightParameter, g_defaultHeight );
+					ConstObjectPtr geo = quad( width, height );
+					NodeAlgo::convert( geo.get(), transformHandle.context(), lightName.c_str() );
```

Comment:
I'm surprised we're getting away with this - this is calling `NSICreate` when we're just meant to be updating an existing node.

Reply:
Ah right, I overlooked that `NodeAlgo::convert()` would call `NSICreate`. I think we may be getting by on a technicality, at least according to the NSI documentation :
```
If the supplied handle matches an existing node, the function does nothing if all
other parameters match the call which created that node.
```

We are creating the node with the same parameters (none) each time. In fact the documentation goes on to say `There are no optional parameters defined as of now.`

I don't know if I'm correct, but either way I should improve how I make geometry updates.

Reply:
Also made obsolete by the improved handling of geometry in https://github.com/GafferHQ/gaffer/pull/5603/commits/a0c61d62e4bb80bef1f2ddfe3839f68389fcb36d.